### DISCUSSION TITLE
Timeline

### DISCUSSION THREAD

**Discussion by gkim132** [jspsych/jsPsych#889] [Q&A] [answered]
Hello,

I changed the first code to the second code in order to add the fixation function. The first code works well, but i got errors on the second one. I feel it should work, but it doesn't.
Any help will be greatly appreciated!
Thank you!
I got this error. 

jspsych.js:759 Trial level node is missing the "type" parameter. The parameters for the node are: {"0":{"timeline":[{"type":"html-keyboard-response","stimulus":"<p class=\"stimulus\">+</p>","choices":"none","trial_duration":500},{"type":"html-keyboard-response","stimulus":"<div>\n                <div style=\"display:inline-block\"><img\n                style=\"width: 30vw;\"\n                src=  t_img_2/5.gif ></div>\n                <div style=\"display:inline-block\"><img \n                style=\"width: 30vw;\"\n                src= t_img_2/5.gif></div>\n                </div>","choices":["f","j"],"data":{"file_name":"t_img_2/5.gif , t_img_2/5.gif","test_part":"test","correct_response":"f"}}]},"1":{"timeline":[{"type":"html-keyboard-response","stimulus":"<p class=\"stimulus\">+</p>","choices":"none","trial_duration":500},{"type":"html-keyboard-response","stimulus":"<div>\n                <div style=\"display:inline-block\"><img\n                style=\"width: 30vw;\"\n                src=  t_img_2/2.gif ></div>\n                <div style=\"display:inline-block\"><img \n                style=\"width: 30vw;\"\n                src= t_img_2/2.gif></div>\n                </div>","choices":["f","j"],"data":{"file_name":"t_img_2/2.gif , t_img_2/2.gif","test_part":"test","correct_response":"f"}}]},"2":{"timeline":[{"type":"html-keyboard-response","stimulus":"<p class=\"stimulus\">+</p>","choices":"none","trial_duration":500},{"type":"html-keyboard-response","stimulus":"<div>\n                <div style=\"display:inline-block\"><img\n                style=\"width: 30vw;\"\n                src=  t_img_2/1.gif ></div>\n                <div style=\"display:inline-block\"><img \n                style=\"width: 30vw;\"\n                src= t_img_2/1.gif></div>\n                </div>","choices":["f","j"],"data":{"file_name":"t_img_2/1.gif , t_img_2/1.gif","test_part":"test","correct_response":"j"}}]},"3":{"timeline":[{"type":"html-keyboard-response","stimulus":"<p class=\"stimulus\">+</p>","choices":"none","trial_duration":500},{"type":"html-keyboard-response","stimulus":"<div>\n                <div style=\"display:inline-block\"><img\n                style=\"width: 30vw;\"\n                src=  t_img_2/4.gif ></div>\n                <div style=\"display:inline-block\"><img \n                style=\"width: 30vw;\"\n                src= t_img_2/4.gif></div>\n                </div>","choices":["f","j"],"data":{"file_name":"t_img_2/4.gif , t_img_2/4.gif","test_part":"test","correct_response":"j"}}]},"4":{"timeline":[{"type":"html-keyboard-response","stimulus":"<p class=\"stimulus\">+</p>","choices":"none","trial_duration":500},{"type":"html-keyboard-response","stimulus":"<div>\n                <div style=\"display:inline-block\"><img\n                style=\"width: 30vw;\"\n                src=  t_img_2/3.gif ></div>\n                <div style=\"display:inline-block\"><img \n                style=\"width: 30vw;\"\n                src= t_img_2/3.gif></div>\n                </div>","choices":["f","j"],"data":{"file_name":"t_img_2/3.gif , t_img_2/3.gif","test_part":"test","correct_response":"f"}}]}}
(anonymous) @ jspsych.js:759
TimelineNode @ jspsych.js:767
(anonymous) @ jspsych.js:751
TimelineNode @ jspsych.js:767
window.jsPsych.core.init @ jspsych.js:160
(anonymous) @ comparision.html:89
jspsych.js:899 Uncaught TypeError: Cannot read property 'info' of undefined
    at evaluateFunctionParameters (jspsych.js:899)
    at doTrial (jspsych.js:834)
    at startExperiment (jspsych.js:784)
    at jspsych.js:2569
    at Object.jsPsych.pluginAPI.module.preloadVideo (jspsych.js:2449)
    at jspsych.js:2568
    at Object.jsPsych.pluginAPI.module.preloadAudioFiles (jspsych.js:2295)
    at jspsych.js:2567
    at Object.jsPsych.pluginAPI.module.preloadImages (jspsych.js:2400)
    at Object.jsPsych.pluginAPI.module.autoPreload (jspsych.js:2566)

<The fisrt code>

`
var images_1 = ["t_img_1/1.gif", "t_img/2.gif","t_img/3_J.gif","t_img/4.gif","t_img/5.gif","t_img/6_J.gif","t_img/7_J.gif","t_img/8_J.gif","t_img/9.gif","t_img/10.gif","t_img/11_J.gif","t_img/12.gif"];
`
`    
var images_2 = ["t_img_2/1.gif", "t_img_2/2.gif","t_img_2/3.gif","t_img_2/4.gif","t_img_2/5.gif","t_img_2/6.gif","t_img_2/7.gif","t_img_2/8.gif","t_img_2/9.gif","t_img_2/10.gif","t_img_2/11.gif","t_img_2/12.gif"];
 `

    var answer_keys = [2,1,1,2,1];

    var trials = [];   
    var random_generation = shuffle(Array.from(Array(answer_keys.length).keys()));
    
    // needed to be simplfied as in a function getting three lists, images_1, images_2, and answer_keys as parameter
    for (var i = 0; i < images_1.length; i++) {
      index = random_generation[i];
      if(answer_keys[index] == 1){
        var trial_data = {file_name :images_1[index] +" , " +images_2[index], test_part: 'test', correct_response: 'f'};
      }
      else if(answer_keys[index] == 2){
        var trial_data = {file_name :images_1[index] +" , " +images_2[index], test_part: 'test', correct_response: 'j'};
      }
      trials.push({
        stimulus: `<div>
                <div style="display:inline-block"><img
                style="width: 30vw;"
                src=  ${ images_1[index] } ></div>
                <div style="display:inline-block"><img 
                style="width: 30vw;"
                src= ${ images_2[index] }></div>
                </div>`,
        data: trial_data            
        
      });
    }
    var test = {
      type: 'html-keyboard-response',
      choices: ['f', 'j'],
      on_finish: function(data){
        data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
      },

      timeline: trials
    }

    jsPsych.init({
      timeline: test,
      //timeline: timeline
      on_finish: function() {
        jsPsych.data.displayData();
      }
    });`

<The second code>

`
var images_1 = ["t_img_1/1.gif", "t_img/2.gif","t_img/3_J.gif","t_img/4.gif","t_img/5.gif","t_img/6_J.gif","t_img/7_J.gif","t_img/8_J.gif","t_img/9.gif","t_img/10.gif","t_img/11_J.gif","t_img/12.gif"];
`
`    
var images_2 = ["t_img_2/1.gif", "t_img_2/2.gif","t_img_2/3.gif","t_img_2/4.gif","t_img_2/5.gif","t_img_2/6.gif","t_img_2/7.gif","t_img_2/8.gif","t_img_2/9.gif","t_img_2/10.gif","t_img_2/11.gif","t_img_2/12.gif"];
 `

    var answer_keys = [2,1,1,2,1];

    var trials = [];   
    var random_generation = shuffle(Array.from(Array(answer_keys.length).keys()));
    
    for (var i = 0; i < images_1.length; i++) {
      index = random_generation[i];
      if(answer_keys[index] == 1){
        var trial_data = {file_name :images_1[index] +" , " +images_2[index], test_part: 'test', correct_response: 'f'};
      }
      else if(answer_keys[index] == 2){
        var trial_data = {file_name :images_1[index] +" , " +images_2[index], test_part: 'test', correct_response: 'j'};
      }
      trials.push({
        timeline:[
          {
            type: 'html-keyboard-response',
            stimulus: '<p class="stimulus">+</p>',
            choices: jsPsych.NO_KEYS,
            trial_duration: 500
          },
          {
            type: 'html-keyboard-response',
            stimulus: `<div>
                <div style="display:inline-block"><img
                style="width: 30vw;"
                src=  ${ images_1[index] } ></div>
                <div style="display:inline-block"><img 
                style="width: 30vw;"
                src= ${ images_2[index] }></div>
                </div>`,
            choices: ['f', 'j'], 
            data: trial_data,
            on_finish: function(data){
              data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
            }
          }        
        ] 
      });
    }        

    jsPsych.init({
      timeline: trials,
      //timeline: timeline
      on_finish: function() {
        jsPsych.data.displayData();
      }
    });` `

**Comment by becky-gilbert** [comment, marked as answer]
Hi @gkim132!
In the second version, every time you call `trials.push()`, you're adding an object with a single `timeline` key:
```javascript
{
   timeline: [ ... ]
}
``` 
but jsPsych doesn't know what to do with that. Instead, you can push your two trial objects directly onto the `trials` array, without putting them inside of an object with a nested timeline. Like this:
```javascript
for (var i = 0; i < images_1.length; i++) {
  index = random_generation[i];
  if(answer_keys[index] == 1){
    var trial_data = {file_name :images_1[index] +" , " +images_2[index], test_part: 'test', correct_response: 'f'};
  }
  else if(answer_keys[index] == 2){
    var trial_data = {file_name :images_1[index] +" , " +images_2[index], test_part: 'test', correct_response: 'j'};
  }
  trials.push(
      {
        type: 'html-keyboard-response',
        stimulus: '<p class="stimulus">+</p>',
        choices: jsPsych.NO_KEYS,
        trial_duration: 500
      },
      {
        type: 'html-keyboard-response',
        stimulus: `<div>
            <div style="display:inline-block"><img
            style="width: 30vw;"
            src=  ${ images_1[index] } ></div>
            <div style="display:inline-block"><img 
            style="width: 30vw;"
            src= ${ images_2[index] }></div>
            </div>`,
        choices: ['f', 'j'], 
        data: trial_data,
        on_finish: function(data){
          data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
        }      
     }
   );
}  
```
Now your `trials` array will consist of a sequence of objects that define the fixations and image response trials. Does that work? And is that what you're trying to do?

**Reply by gkim132** [reply]
Hi @becky-gilbert ,

Thank you for your reply.
Yes, I try to add fixation right before each trial.
However, unfortunately, it doesn't work.  :(

**Reply by gkim132** [reply]
I still have the same errors:
jspsych.js:759 Trial level node is missing the "type" parameter. The parameters for the node are: {"0":{"type":"html-keyboard-response","stimulus":"<p class=\"stimulus\">+</p>","choices":"none","trial_duration":500},"1":{"type":"html-keyboard-response","stimulus":"<div>\n                <div style=\"display:inline-block\"><img\n                style=\"width: 30vw;\"\n                src=  t_img_2/5.gif ></div>\n                <div style=\"display:inline-block\"><img \n                style=\"width: 30vw;\"\n                src= t_img_2/5.gif></div>\n                </div>","choices":["f","j"],"data":{"file_name":"t_img_2/5.gif , t_img_2/5.gif","test_part":"test","correct_response":"f"}},"2":{"type":"html-keyboard-response","stimulus":"<p class=\"stimulus\">+</p>","choices":"none","trial_duration":500},"3":{"type":"html-keyboard-response","stimulus":"<div>\n                <div style=\"display:inline-block\"><img\n                style=\"width: 30vw;\"\n                src=  t_img_2/1.gif ></div>\n                <div style=\"display:inline-block\"><img \n                style=\"width: 30vw;\"\n                src= t_img_2/1.gif></div>\n                </div>","choices":["f","j"],"data":{"file_name":"t_img_2/1.gif , t_img_2/1.gif","test_part":"test","correct_response":"j"}},"4":{"type":"html-keyboard-response","stimulus":"<p class=\"stimulus\">+</p>","choices":"none","trial_duration":500},"5":{"type":"html-keyboard-response","stimulus":"<div>\n                <div style=\"display:inline-block\"><img\n                style=\"width: 30vw;\"\n                src=  t_img_2/2.gif ></div>\n                <div style=\"display:inline-block\"><img \n                style=\"width: 30vw;\"\n                src= t_img_2/2.gif></div>\n                </div>","choices":["f","j"],"data":{"file_name":"t_img_2/2.gif , t_img_2/2.gif","test_part":"test","correct_response":"f"}},"6":{"type":"html-keyboard-response","stimulus":"<p class=\"stimulus\">+</p>","choices":"none","trial_duration":500},"7":{"type":"html-keyboard-response","stimulus":"<div>\n                <div style=\"display:inline-block\"><img\n                style=\"width: 30vw;\"\n                src=  t_img_2/4.gif ></div>\n                <div style=\"display:inline-block\"><img \n                style=\"width: 30vw;\"\n                src= t_img_2/4.gif></div>\n                </div>","choices":["f","j"],"data":{"file_name":"t_img_2/4.gif , t_img_2/4.gif","test_part":"test","correct_response":"j"}},"8":{"type":"html-keyboard-response","stimulus":"<p class=\"stimulus\">+</p>","choices":"none","trial_duration":500},"9":{"type":"html-keyboard-response","stimulus":"<div>\n                <div style=\"display:inline-block\"><img\n                style=\"width: 30vw;\"\n                src=  t_img_2/3.gif ></div>\n                <div style=\"display:inline-block\"><img \n                style=\"width: 30vw;\"\n                src= t_img_2/3.gif></div>\n                </div>","choices":["f","j"],"data":{"file_name":"t_img_2/3.gif , t_img_2/3.gif","test_part":"test","correct_response":"f"}}}
(anonymous) @ jspsych.js:759
TimelineNode @ jspsych.js:767
(anonymous) @ jspsych.js:751
TimelineNode @ jspsych.js:767
window.jsPsych.core.init @ jspsych.js:160
(anonymous) @ comparision.html:87
jspsych.js:899 Uncaught TypeError: Cannot read property 'info' of undefined
    at evaluateFunctionParameters (jspsych.js:899)
    at doTrial (jspsych.js:834)
    at startExperiment (jspsych.js:784)
    at jspsych.js:2569
    at Object.jsPsych.pluginAPI.module.preloadVideo (jspsych.js:2449)
    at jspsych.js:2568
    at Object.jsPsych.pluginAPI.module.preloadAudioFiles (jspsych.js:2295)
    at jspsych.js:2567
    at Object.jsPsych.pluginAPI.module.preloadImages (jspsych.js:2400)
    at Object.jsPsych.pluginAPI.module.autoPreload (jspsych.js:2566)
evaluateFunctionParameters @ jspsych.js:899
doTrial @ jspsych.js:834
startExperiment @ jspsych.js:784
(anonymous) @ jspsych.js:2569
jsPsych.pluginAPI.module.preloadVideo @ jspsych.js:2449
(anonymous) @ jspsych.js:2568
jsPsych.pluginAPI.module.preloadAudioFiles @ jspsych.js:2295
(anonymous) @ jspsych.js:2567
jsPsych.pluginAPI.module.preloadImages @ jspsych.js:2400
jsPsych.pluginAPI.module.autoPreload @ jspsych.js:2566
(anonymous) @ jspsych.js:184
checkExclusions @ jspsych.js:997
window.jsPsych.core.init @ jspsych.js:179
(anonymous) @ comparision.html:87

**Reply by jodeleeuw** [reply]
Hi @gkim132. Can you post the code that you are currently using?

**Reply by gkim132** [reply]
Hi @jodeleeuw,

Sure! here is the code.
Please help me out with these errors.
Thank you!
```
<!DOCTYPE html>
<html>
  <head>
    <title>Comparison</title>
    <script src="jspsych-6.1.0/jspsych.js"></script>
    <script src="jspsych-6.1.0/plugins/jspsych-html-keyboard-response.js"></script>
    <script src="jspsych-6.1.0/plugins/jspsych-image-keyboard-response.js"></script>
    <script src="jspsych-6.1.0/plugins/jspsych-instructions.js"></script>
    <script src="jspsych-6.1.0/plugins/jspsych-fullscreen.js"></script>
    <link href="jspsych-6.1.0/css/jspsych.css" rel="stylesheet" type="text/css"></link>
  </head>
  <body></body>
  

  <script>
    
    //A triplet consists of the n_th element of each array (images_1, images_2, answer_keys) for n=1,2,3,...
    //In each triplet, if the element in the list "answer_keys" is x, then the element in "images_x" is the correct answer
    //For example, the first element in images_2 will be the answer since the first element of answer_keys is 2.
    //Images_1 will be located on the left of the screen
    var images_1 = [ "t_img_2/1.gif", "t_img_2/2.gif","t_img_2/3.gif","t_img_2/4.gif","t_img_2/5.gif"];
    
    //Images_2 will be located on the right of the screen
    var images_2 = [ "t_img_2/1.gif", "t_img_2/2.gif","t_img_2/3.gif","t_img_2/4.gif","t_img_2/5.gif"];
    //The subject would want to press "f" when answer_keys is 1 and "j" otherwise.
    var answer_keys = [2,1,1,2,1];
    var trials = [];   
    var random_generation = shuffle(Array.from(Array(answer_keys.length).keys()));
    var fixation = {
      type: 'html-keyboard-response',
      stimulus: '<div style="font-size:60px;">+</div>',
      choices: jsPsych.NO_KEYS,
      trial_duration: 1000
    }

    for (var i = 0; i < images_1.length; i++) {
      index = random_generation[i];
      if(answer_keys[index] == 1){
        var trial_data = {file_name :images_1[index] +" , " +images_2[index], test_part: 'test', correct_response: 'f'};
      }
      else if(answer_keys[index] == 2){
        var trial_data = {file_name :images_1[index] +" , " +images_2[index], test_part: 'test', correct_response: 'j'};
      }
      trials.push({
        timeline: [
          {
            type: 'html-keyboard-response',
            stimulus: '<div style="font-size:60px;">+</div>',
            choices: jsPsych.NO_KEYS,
            trial_duration: 1000
          },
          {
            type: 'html-keyboard-response',
            stimulus: `<div>
                    <div style="display:inline-block"><img
                    style="width: 30vw;"
                    src=  ${ images_1[index] } ></div>
                    <div style="display:inline-block"><img 
                    style="width: 30vw;"
                    src= ${ images_2[index] }></div>
                    </div>`,
            choices: ['f', 'j'],
            data: trial_data,
            on_finish: function(data){
              data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
            }
          }
        ]      
      });
    }    

    var end_message = {
      type: 'html-keyboard-response',
      stimulus: '<p>End</p>',
    };
    
   
    jsPsych.init({
      timeline: [ trials, end_message],
      on_finish: function() {
        jsPsych.data.displayData();
      }
    });
    //function for shuffle
    function shuffle(a) {
        for (let i = a.length - 1; i > 0; i--) {
            const j = Math.floor(Math.random() * (i + 1));
            [a[i], a[j]] = [a[j], a[i]];
        }
        return a;
    }
    
  </script>

</html>
```

**Reply by becky-gilbert** [reply]
Hi @gkim132! It looks like this code has the same problem that I mentioned above. Did you try removing the object with the timeline key from your `trial.push()` argument inside the for loop? Doing it this way works for me:

```javascript
trials.push(
    {
        type: 'html-keyboard-response',
        stimulus: '<div style="font-size:60px;">+</div>',
        choices: jsPsych.NO_KEYS,
        trial_duration: 1000
    },
    {
        type: 'html-keyboard-response',
        stimulus: '<div>test '+i+'</div>',
        choices: ['f', 'j'],
        data: trial_data,
        on_finish: function(data){
            data.correct = data.key_press == jsPsych.pluginAPI.convertKeyCharacterToKeyCode(data.correct_response);
        }
    }
);
```

Another problem is that you're creating a timeline array called `trials` but then putting that array inside of another array when you call jsPsych.init(). This means that your main experiment timeline contains two elements, one of which is an array (`trials`). 
Instead, each element in your timeline array should be an _object_ that either defines a trial or some kind of trial procedure (using a timeline along with timeline variables, conditional functions and/or loop functions). 
One way to fix this is to push your `end_message` trial object onto your `trials` array, and use the `trials` array as your main experiment timeline:
```javascript
trials.push(end_message);

jsPsych.init({
    timeline: trials, 
    ...
```

One last thing to note is that I think you'll want to manually preload your images using the preload_images parameter in jsPsych.init(), because you're embedding the images in HTML strings (rather than using one of the jspsych-image-* plugins). 
```javascript
var all_images = images_1.concat(images_2);

jsPsych.init({
    timeline: trials,
    preload_images: all_images,
    ...
```

I hope this helps!

**Reply by gkim132** [reply]
Hi @becky-gilbert,
Thank you so much for your reply. 
It works now.  Although I tried removing the object with the timeline key from my trial.push(), there were the same errors. But, it's been fixed with your solution in the second paragraph. 
The detailed explanation of why I got the errors is very helpful. Thank you.

What is the purpose of using the preload_images parameter?

**Reply by becky-gilbert** [reply]
Great! And sorry that I hadn't noticed the issue with your jsPsych.init timeline when I responded the first time.

The preload_images parameter will load your image files into the participant's browser before the experiment starts. You can read more about preloading in the jsPsych docs [here](https://www.jspsych.org/overview/media-preloading/).

**Reply by gkim132** [reply]
Thank you so much @becky-gilbert !
I've learned a lot!
